### ISSUE TITLE
Add support for custom shield shapes

### ISSUE THREAD

**Issue by CringeStar** [StellarWind22/Fabric-Shield-Lib#158]
I think this would be a big feature, but it would be very beneficial. I think we would have to look into the shield model, the texture, the rendering method, and the banner pattern textures.

**Comment by CringeStar** [comment]
The custom shape has been successful, though it is very manual so I don't know how much the library can help with that. A bigger obstacle is the banner designs, as for every custom shape of shield, a new set would have to be created and somehow be used instead of the default vanilla. Simply overwriting the vanilla textures doesn't work too well, even though that is the only way I can find of doing so. Overwriting the textures breaks the vanilla shield's pattern and it limits the mod creator to only 1 custom-shaped shield.

**Comment by CringeStar** [comment]
![2023-09-27_16 46 22](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/4c742617-10fd-4532-9c6e-31ae7ea0fc61)
![2023-09-27_16 46 23](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/4a333f4d-466c-4d73-bd4c-35dc4242fdce)
![2023-09-27_16 46 26](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/ee71e113-5c88-416a-b3f0-f14dbe3a24de)
![2023-09-27_16 46 28](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/f8796927-1d11-4559-989b-c3b4edf3a9ca)
![2023-09-27_18 04 53](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/1f3d6570-4759-43e2-a473-5ca46c325412)
![2023-09-27_18 04 55](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/1f8de921-ca56-41d3-99ac-36ea5bc48677)
![2023-09-27_18 04 57](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/dc692ae5-aabf-4842-8bc8-0983e4618918)
![2023-09-27_18 05 12](https://github.com/StellarWind22/Fabric-Shield-Lib/assets/87676346/7272a045-cb9d-4918-838e-b158646cf2eb)

**Comment by CringeStar** [comment]
Additionally, the custom looks weird in the hand opposite of which it was designed for. For example, I designed this test shape when the shield was in the main hand, so it looks weird in the offhand.

**Comment by StellarWind22** [comment]
We would need to write our own banner rendering system that can take in custom sets of banner textures and render them on the shield

**Comment by StellarWind22** [comment]
for every new shield shape the end user would need to add a new set of banner textures to match

**Comment by CringeStar** [comment]
Yeah that was the biggest issue I faced. I couldn't find how to direct rendering for the custom shield to a custom set of banner patterns. I had to replace the vanilla textures. I think I attempted a mixin to change the path, but I'll have to look at it again.

**Comment by CringeStar** [comment]
In 1.20.5/6 there was an update to banner patterns for shield, it seems it is easier to add custom ones. Will look into it over the summer.

**Comment by CringeStar** [comment]
Maybe look into DataComponentTypes.BANNER_PATTERNS

**Comment by StellarWind22** [comment]
> Yeah I've been working on this. I'm creating an AbstractShieldModel class where the end user defines the getPlate and getHandle methods & also some uv stuff to tell a custom banner renderer where to draw the banner texture on the shield.

Hmmm. I wonder if it should be an interface instead so you don't have to rely on class inheritance?
